decisions/2024-08-30_central_sync_queue_pull_reduction.md

```diff
@@ -0,0 +1,164 @@
+
+But also would work. We could make master list queued data again and improving queuing with filtering to only send updates to where they need to go. This would fix the initialisation problem and also improve legacy mSupply performance.
+
+_Pros:_
+
+- Should improve sync performance over all, including UX on legacy central servers for users doing admin. (There is merit to fixing the master list queuing filtering on LCS regardless of where this KDD goes!)
+- Fixes the problem for existing sites all versions of OMS and legacy apps.
+
+_Cons:_
+
+- Still potentially queuing lots of records to many sites. Some users will have >1000 sites and a general master list used across them with >5000 items. These instances will still suffer millions of queued records that central_change_log set out to fix to begin with!
+- Eventually we want OMS only deployments where OMS controls central data, so a solution in LCS land would require us eventually fixing it again in OMCS land.
+- Doing work in LCS when we're trying to focus on OMS
+
+### Option 2 - LCS central_change_log filtering per log processed
+
+We implement a system that allows sync v5 change log requests to have some filtering on the LCS side.
+
+Suppose a site requests 5000 change_log records from cursor 0:
+
+1. If the change_log record is for a list_master_line,
+2. Get all the list_master.ID values for lists that are visible to the site; that is from list_master_name_join for all stores (this can be done once for the request and reused in the process, or cached between requests to reduce querying overheads)
+3. Get the list_master_line record
+4. If the list_master_line.master_list_id is **not** in that set, skip this change log record.
+
+_Pros:_
+
+- Stay in changelog land and don't queue loads of records
+
+_Cons:_
+
+- Still looks like 90k records to pull, even if large amounts get filtered out.
+- If 5000 records are requested and we filter out some (or all) of them, do we scan ahead to grab some more records till the batch size is fulfilled (and continuing until done)? Probably negligible but none the less a slower.
+- Added overhead to getting records
+- If they actually need all of the master list lines, it's more overhead doing filtering for no reason. Can probably optimise this case though.
+- Doing work in LCS when we're trying to focus on OMS
+- Eventually we want OMS only deployments where OMS controls central data, so a solution in LCS land would require us eventually fixing it again in OMCS land.
+- Only fixes for OMS
+
+### Option 3 - OMCS change_log filtering per log processed
+
+The filtering could be more or less the same as in Option 2, but we do it in OMCS. Go full hog: We change to OMCS changelog pull and drop pulling LCS changelog.
```

Comment:
In general I like this idea, especially as a bunch of R&D Work has been done on this possibility.
Then we're developing in the OMS code base rather than more code in legacy.